File: 2023/day01.js
const fs = require('fs');
const input = fs.readFileSync('input/day01', 'utf8');
const lines = input.trim().split('\n');
const calibrationValue = line => Number(line[0] + line.slice(-1));

const wordsToDigits = (line) => {
  if (line == '') return '';
  if (line.startsWith('one')) return '1' + wordsToDigits(line.slice(2));
  if (line.startsWith('two')) return '2' + wordsToDigits(line.slice(2));
  if (line.startsWith('three')) return '3' + wordsToDigits(line.slice(4));
  if (line.startsWith('four')) return '4' + wordsToDigits(line.slice(4));
  if (line.startsWith('five')) return '5' + wordsToDigits(line.slice(3));
  if (line.startsWith('six')) return '6' + wordsToDigits(line.slice(3));
  if (line.startsWith('seven')) return '7' + wordsToDigits(line.slice(4));
  if (line.startsWith('eight')) return '8' + wordsToDigits(line.slice(4));
  if (line.startsWith('nine')) return '9' + wordsToDigits(line.slice(3));
  if (/^\d/.test(line)) return line[0] + wordsToDigits(line.slice(1));
  return wordsToDigits(line.slice(1));
};

console.log(`Part 1: ${lines.reduce((acc, line) => acc + calibrationValue(line.replace(/\D/g, '')), 0)}`);
console.log(`Part 2: ${lines.reduce((acc, line) => acc + calibrationValue(wordsToDigits(line)), 0)}`);

File: 2023/day01.py
import re

def calibration_value(line):
    return int(line[0] + line[-1])

def words_to_digits(line):
    if line == '': return ''
    if line.startswith('one'): return '1' + words_to_digits(line[2:])
    if line.startswith('two'): return '2' + words_to_digits(line[2:])
    if line.startswith('three'): return '3' + words_to_digits(line[4:])
    if line.startswith('four'): return '4' + words_to_digits(line[4:])
    if line.startswith('five'): return '5' + words_to_digits(line[3:])
    if line.startswith('six'): return '6' + words_to_digits(line[3:])
    if line.startswith('seven'): return '7' + words_to_digits(line[4:])
    if line.startswith('eight'): return '8' + words_to_digits(line[4:])
    if line.startswith('nine'): return '9' + words_to_digits(line[3:])
    if line[0].isdigit(): return line[0] + words_to_digits(line[1:])
    return words_to_digits(line[1:])

calibration_data = open('input/day01', 'r').read().splitlines()
print('Part 1: ', sum(map(calibration_value, map(lambda x: re.sub(r'\D', '', x), calibration_data))))
print('Part 2: ', sum(map(calibration_value, map(words_to_digits, calibration_data))))
